### PR TITLE
ECS cluster template on spot and on-demand instances

### PR DESCRIPTION
**Pull request by IYermakov**
## Github Issue #

https://github.com/IYermakov/DevOpsA3Training/issues/32

## What Are We Doing Here?

PR adds a new template according to work on new kind of deployment to ECS cluster with EC2 launch type.

## How to Test and Verify

1. Check out this PR
2. Follow the instructions from README file

### Level

Medium

### Required Testing

`Smoke`, `Manual`

### Risk Summary

It introduces a part of new functionality but doesn't break any previous work.

## Code Review Checklist

- [ ] Testing instructions are included. If not, make sure you leave us a line or two for the reason.

- [ ] Well documented, including updates to any necessary markdown files. When
we inevitably come back to this code it will only take hours to figure out, not
days.

- [ ] Make sure PR has only the changes required for the Jira issues.

- [ ] Have you make sure to follow the agreements in our process document?

- [ ] Have someone from a different region than you review your changes. If not, make sure to leave a comment on why it is not required

- [ ] If it's applicable, document the deployment plan for the changes.

### CODE REVIEW

File: ops/cloudformation/ECS/ecs-cluster.yml [hunk 1]
```diff
@@ -0,0 +1,370 @@
...
+    Description: Minimum value of EC2 instances in cluster.
+
+  MaxSize:
+    Type: String
+    Default: 10
+    Description: Maximum value of EC2 instances in cluster.
+
+  ECSAMI:
+    Type: AWS::EC2::Image::Id
+    Default: ami-097e3d1cdb541f43e
```
Comment (mike-unitskyi) [inline ops/cloudformation/ECS/ecs-cluster.yml]: remove `Default` as it's region/account specific

File: ops/cloudformation/ECS/ecs-cluster.yml [hunk 2]
```diff
@@ -0,0 +1,370 @@
...
+    Description: The name of the parent networking stack that you created.
+
+  ALBStackName:
+    Type: String
+    Default: alb
+    Description: The name of the previously created ALB stack.
+
+  Environment:
+    Type: String
+    Default: dev
```
Comment (mike-unitskyi) [inline ops/cloudformation/ECS/ecs-cluster.yml]: remove default

File: ops/cloudformation/ECS/ecs-cluster.yml [hunk 3]
```diff
@@ -0,0 +1,370 @@
...
+Description: "Stack creates ECS cluster and roles, ASG for container instances."
+
+Parameters:
+
+  KeyPairName:
+    Type: AWS::EC2::KeyPair::KeyName
+    Description: Name of the SSH keypair to use for the EC2 instances.
+    Default: devopsa3
+
+  VPCStackName:
```
Comment (mike-unitskyi) [inline ops/cloudformation/ECS/ecs-cluster.yml]: same comment for extracting into map

File: ops/cloudformation/ECS/ecs-cluster.yml [hunk 4]
```diff
@@ -0,0 +1,370 @@
...
+                - 'logs:PutLogEvents'
+              Resource: '*'
+
+Outputs:
+
+  ClusterName:
+    Description: The name of the ECS cluster
+    Value: !Ref 'ECSCluster'
+    Export:
+      Name: !Join [ ':', [ !Ref 'AWS::StackName', 'ClusterName' ] ]
```
Comment (mike-unitskyi) [inline ops/cloudformation/ECS/ecs-cluster.yml]: here and below use `!Sub`

File: ops/cloudformation/ECS/ecs-cluster.yml [hunk 5]
```diff
@@ -0,0 +1,370 @@
...
+        SecurityGroupIds: [!Ref 'EcsHostSecurityGroup']
+        TagSpecifications:
+        - ResourceType: instance
+          Tags:
+          - Key: Name
+            Value: !Ref AWS::StackName
+          - Key: Environment
+            Value: !Ref Environment
+        UserData:
+          Fn::Base64: !Sub |
```
Comment (mike-unitskyi) [inline ops/cloudformation/ECS/ecs-cluster.yml]: could you describe how app version/docker image is specified? or it's out of scope of this PR?

Comment (IYermakov) [reply ops/cloudformation/ECS/ecs-cluster.yml]: yep, it's out of scope

Comment (mike-unitskyi) [reply ops/cloudformation/ECS/ecs-cluster.yml]: add comment and link to issue

File: ops/cloudformation/ECS/ecs-cluster.yml (hunk 1 above)
Comment (IYermakov) [reply ops/cloudformation/ECS/ecs-cluster.yml]: removed

File: ops/cloudformation/ECS/ecs-cluster.yml (hunk 2 above)
Comment (IYermakov) [reply ops/cloudformation/ECS/ecs-cluster.yml]: removed

File: ops/cloudformation/ECS/ecs-cluster.yml (hunk 5 above)
Comment (IYermakov) [reply ops/cloudformation/ECS/ecs-cluster.yml]: https://github.com/IYermakov/DevOpsA3Training/issues/31

Comment (mike-unitskyi) [reply ops/cloudformation/ECS/ecs-cluster.yml]: it should be added to code

File: ops/cloudformation/ECS/ecs-cluster.yml (hunk 4 above)
Comment (IYermakov) [reply ops/cloudformation/ECS/ecs-cluster.yml]: Okay, done

File: ops/cloudformation/ECS/ecs-cluster.yml (hunk 3 above)
Comment (mike-unitskyi) [reply ops/cloudformation/ECS/ecs-cluster.yml]: issue for this?

Comment (IYermakov) [reply ops/cloudformation/ECS/ecs-cluster.yml]: here it is - https://github.com/IYermakov/DevOpsA3Training/issues/29

**Pull request merged**